### PR TITLE
APM implementation based on uniswap lp token mechanism. 

### PR DESCRIPTION
**Pull request by toufic0710**
V0 for the apm. @dr-chain 
Restrictions for update functions of apm still need to be done.
gas testing : get reserve 37k, updateAddLiquidity 57k .

### CODE REVIEW

**Comment by Edoumou** [comment]
- Variable must be named properly, this is important when reviewing the contract and for audit. As an example, on line 29, a variable named **addressesToPair** is defined as a mapping of **Pair** structure. so the name here doesn't reflect the object, making the review difficult. Good practise if a remember is:

```solidity
  struct Person {
    string name;
    uint256 age;
  }

 mapping(string => Person) person;
```
- Why there is a constructor defining the reserves (initial reserve if I got it) ? This doesn't reflect scenario we are trying to implement. The function that adds liquidity must be able to handle also the case of first deposit.

- I can't see any function that allows to create new pools. What happens when a new pair has been approved by the Gov. contract and that someone would like to add Liquidity to that pair (initial deposit) ? Which function in this APM version should he call ?

- In the constructor, lines **37** and **40** there are two instances of same pair. There is only one pair from two addresses, therefore, when the contract "architecture" is well defined, there should only be one instance of Pair struct. In this APMP version it is considered that there are two pairs from two addresses, which in my opinion reflect a misunderstanding of what a pair should be. 

- As for variables that have been defined:

```solidity
    mapping(address => uint256) internal totalReserve;
    mapping(address => uint256) internal totalVlp;
    mapping(address => mapping( address => Pair) ) addressesToPair;

    struct Pair {
      uint vlpTokenA;
      uint vlpTokenB;
    }
```

What is totalVlp ? how is it different from vlpTokenA and vlpTokenB ? If this variable is related to a pair, it should be part of the Pair structure. Maybe an explanation will make this part more clear.

- In **getReserves** function:

```solidity
  (reserveA, reserveB) = (vlpA * totalReserve[tokenA] / totalVlpA, vlpB * totalReserve[tokenB] / totalVlpB);
```

There is a structure Pair that defines a pair, and that contains two fields: **vlpTokenA** and **vlpTokenB**. Why these two variables are not considered as reserve of each token in that pool ? 

Here is Uniswapp **getReserves** function:

```solidity
  function getReserves() public view returns (uint112 _reserve0, uint112 _reserve1, uint32 _blockTimestampLast) {
        _reserve0 = reserve0;
        _reserve1 = reserve1;
        _blockTimestampLast = blockTimestampLast;
    }
```

It returns the amount of tokens in the pair, which for this APM version should be **vlpTokenA** and **vlpTokenB**.

So it is not so clear as for the role of each variable. Probably need more details and explanation about all variables.

In the same function, one can see the following lines of code:

```solidity
  Pair storage pair = addressesToPair[token0][token1];
  (uint vlp0, uint vlp1) = (pair.vlpTokenA, pair.vlpTokenB);
  (uint vlpA, uint vlpB) = tokenA == token0 ? (vlp0, vlp1) : (vlp1, vlp0);
```

I don't understand why the second line has been added since the following gives exactly same result

```solidity
  tokenA == token0 ?
    (uint vlpA, uint vlpB) = (pair.vlpTokenA, pair.vlpTokenB):
    (uint vlpA, uint vlpB) = (pair.vlpTokenB, pair.vlpTokenA);
```

Always avoid unnecessary instructions. Happy here it's just a view function, if these kind of intermediate steps are used in a function that alter state variables, then it should trigger unnecessary gas.

- In ** getVlps** function: same remark as previously.

- In function ** updateAddLiquidity**:

Also here, the name of the function is confusing. What does it mean update and add liquidity ? either liquidity is added or reserve are updated.

To update the two fields of the pair struct, the function **amountToAddVlp** is called. The calculation is similar than using a weight:

```solidity
  function amountToAddVlp(uint oldVlp, uint amount, uint totalVlpToken, uint totalReserveToken) public pure returns (uint newVlp) {
        newVlp = oldVlp + amount * totalVlpToken / totalReserveToken;
    }
```

so, this function should probably update the reserve properly.

To update **totalVlp** the function **updateTotalVlp** is called. How calculation is done is not clear, so need explanation.

```solidity
  function updateTotalVlp(uint oldTotalVlp, uint oldVlp, uint newVlp) public pure returns (uint totalVlpToken) {
        totalVlpToken = oldTotalVlp - oldVlp + newVlp;
    }
```

In same function, lines 99: 
```solidity
  else if(totalReserveA == 0 && totalReserveB >0)
```
This can be a case where new pair is added with one of the tokens just being approved by Gov. contract. The __if__, __else if__ and __else__ statement can be avoided some refactoring. 

- **updateRemoveLiquidity**: Here also, the name of the function doesn't reflect what function does.

Also, when you compare this function to **updateAddLiquidity** it is clear that some logic is missing. Here is the **updateRemoveLiquidity**

```solidity
  function updateRemoveLiquidity(
        uint amount, //amountA is the amount of tokenA removed in total pool reserve ( so not the total amount of tokenA in total pool reserve)
        address token) external {

        totalReserve[token] -= amount;
    }
```

However, there are a lot of logic added in **updateAddLiquidity** (30 lines), and since the **updateRemoveLiquidity** function is supposed to have same logic but in the reverse way, then we accept the function to have more instruction than just 

```solidity
  totalReserve[token] -= amount;
```

- Function **updateSwap**

Apart from variable naming which can make things confusing, the formula derived from the constant product is used, the result should be what is expected.

**Comment by Edoumou** [comment]
@GrandGarcon I just added you as a reviewer in order to have at least two people for this review, as my view alone will not suffice.

@dr-chain

**Comment by dr-chain** [comment]
@Edoumou Can you also check what is the difference between the formula you explained to me on Monday evening and this one?

@yuliu-debond had suggested these formula himself.

**Comment by Edoumou** [comment]
@dr-chain I didn't say that the formula is not correct, I cannot guess a formula from a code. What I meant is to tell which formula is used and explain how it is derived.

I also added @GrandGarcon as reviewer in order to have different point of view.

**Comment by dr-chain** [comment]
@Edoumou @toufic0710 We need to merge at this point. @Edoumou can you alias with @toufic0710 if the formula is not clear for you?

**Comment by toufic0710** [comment]
thanks @Edoumou for the review. Here is a version with some minor changes. For all your questions about logic and variable explanations, I was writing it on the overleaf but @dr-chain said it was not priority for now, so I guess this will be left for later (after submitting to the audit team maybe).

File: contracts/APM.sol [hunk 1]
```diff
@@ -17,106 +17,190 @@ pragma solidity 0.8.13;
 //import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import "./interfaces/IAPM.sol";
 import "./libraries/SafeMath.sol";
+import "./libraries/CDP.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
```
Comment (Edoumou) [inline contracts/APM.sol:23]: Done already

**Review by GrandGarcon — APPROVED** [review]

**Review by Edoumou — APPROVED** [review]

**Pull request merged**